File: src/mat_comp_order.cpp
#include <iostream> // For std::cout, std::cerr, std::endl
#include <vector> // For std::vector
#include <string> // For std::string
#include <map> // For std::map
#include <limits> // For std::numeric_limits
#include <utility> // For std::pair
#include <cstdlib> // For std::atoi

using namespace std;

map<pair<int, int>, pair<int, string>> memo;

pair <int, string> mat_mul_order(const vector <pair<int, int>>& dims, int i, int j) {
    pair <int, int> key {i, j};
    if (memo.find(key) != memo.end()) {
        return memo[key];
    }

    if (i == j) {
        return {0, "M" + to_string(i)};
    }
    
    int res_ops = numeric_limits<int>::max();
    string res_order;
    for (int k = i; k < j; k++) {
        pair <int, string> left = mat_mul_order(dims, i, k);
        pair <int, string> right = mat_mul_order(dims, k + 1, j);
        int cur_ops = dims[i].first*dims[k].second*dims[j].second;
        int total_ops = cur_ops + left.first + right.first;
        if (total_ops < res_ops) {
            res_ops = total_ops;
            res_order = "(" + left.second + " x " + right.second + ")";
        }

    }
    memo[key] = {res_ops, res_order};
    return {res_ops, res_order};


}

int main(int argc, char *argv[]) {
    if (argc < 3 || argc % 2 == 0) {
        cerr << "Usage: " << argv[0] << "<dim1> <dim2> ..." << endl;
    }
    
    vector<pair<int, int>> dims;
    for (int i = 1; i< argc; i+=2) {
        int dim1 = atoi(argv[i]);
        int dim2 = atoi(argv[i+1]);
        dims.push_back({dim1, dim2});
    }
    auto res = mat_mul_order(dims, 0, dims.size()-1);
    cout << "Minimum number of operations: " << res.first << endl;
    cout << "order of multiplication: " << res.second << endl;
    return 0;
}

File: src/mat_comp_order.py
import numpy as np

def mat_mul_order(dims):
    memo = {}

    def recurse(i, j):
        key = (i, j)
        if key in memo:
            return memo[key]
        if i == j:
            return 0, f'M{i}'
        res = np.inf
        order = ""
        for k in range(i, j):
            left_ops, left_order = recurse(i, k)
            right_ops, right_order = recurse(k + 1, j)
            cur_ops = dims[i][0] * dims[k][1] * dims[j][1]
            total_ops = left_ops + right_ops + cur_ops

            if total_ops < res:
                res = total_ops
                order = f'({left_order} x {right_order})'
        
        memo[key] = (res, order)
        return res, order

    min_ops, optimal_order = recurse(0, len(dims) - 1)
    return min_ops, optimal_order

# Example usage
dims = [(2, 3), (3, 4), (4, 2)]
min_ops, optimal_order = mat_mul_order(dims)
print(f"Minimum number of operations: {min_ops}")
print(f"Optimal multiplication order: {optimal_order}")